### PR TITLE
#180 add WatermarkEstimator for CommitLogRead

### PR DESCRIPTION
**Pull request by je-ik**

### CODE REVIEW

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/OffsetRestrictionTracker.java [hunk 1]
```diff
@@ -158,6 +184,29 @@ public int hashCode() {
...
+    OffsetWatermarkEstimator(OffsetRange offsetRange) {
+      this.processedRange = offsetRange;
+    }
+
+    @Override
+    public Instant currentWatermark() {
+      if (processedRange.getEndOffsetInclusive() != null) {
+        return Instant.ofEpochMilli(processedRange.getEndOffsetInclusive().getWatermark());
+      }
+      return BoundedWindow.TIMESTAMP_MIN_VALUE;
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/OffsetRestrictionTracker.java:201]: start offset?

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java [hunk 2]
```diff
@@ -466,16 +465,23 @@ private void handleFlushDataBaseOnPosition(
...
       onIdleRef.set(onIdleFuture);
+      AtomicReference<StreamElement> lastConsumed = new AtomicReference<>();
       BiConsumer<StreamElement, OffsetCommitter> consumer =
           (el, committer) -> {
             try {
               if (!killSwitch.get()) {
                 synchronized (observer) {
                   el = cloneAndUpdateAttribute(getEntityDescriptor(), el);
                   watermark.update(el);
+                  Optional.ofNullable(lastConsumed.get())
```
Comment (dmvk) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:476]: why do we need this - "mark as consumed on next element" approach?

**Review by dmvk — APPROVED** [review]
LGTM 👍  Good job 🎉

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java (hunk 2 above)
Comment (je-ik) [reply direct/core/src/test/java/cz/o2/proxima/direct/storage/InMemStorage.java:476]: we need that to be sure that `observeBulkOffsets` will return the last consumed element (representing the offset).

**Pull request merged**